cstar/roms/simulation.py

```diff
@@ -439,57 +514,165 @@ def codebases(self) -> list[ExternalCodeBase]:
         return [self.codebase, self.marbl_codebase]
 
     @property
-    def in_file(self) -> Path:
-        """Retrieves the ROMS runtime input file (.in) associated with this simulation.
+    def _forcing_paths(self) -> list[Path]:
+        """Collect and return all local paths to ROMS forcing input datasets.
+
+        This internal property gathers the `working_path` attributes of all forcing-related
+        datasets attached to the simulation — including tidal, river, surface, boundary, and
+        correction datasets — and returns a flat list of resolved file paths.
+
+        Returns
+        -------
+        list of pathlib.Path
+            A list of paths to all local forcing input files required by ROMS.
+
+        Raises
+        ------
+        ValueError
+            If any required forcing dataset does not have a `working_path` set,
+            indicating it has not been retrieved or prepared.
+
+        Notes
+        -----
+        - If any `working_path` is a list of files (e.g., multiple NetCDFs), each path
+          is included individually.
+        - This property is used to populate `runtime_settings.forcing`.
+        """
+
+        forcing_sources = chain(
+            [self.tidal_forcing, self.river_forcing],
+            self.surface_forcing,
+            self.boundary_forcing,
+            self.forcing_corrections,
+        )
+
+        forcing_paths: list[Path] = []
+
+        for source in filter(None, forcing_sources):
+            paths = source.working_path
+            if paths is None:
+                raise ValueError(
+                    f"{source.__class__.__name__} does not have "
+                    + "a local working_path. Call ROMSSimulation.setup() or "
+                    + f"{source.__class__.__name__}.get() and try again."
+                )
+
+            elif isinstance(paths, list):
+                forcing_paths.extend(paths)
+            else:
+                forcing_paths.append(paths)
+
+        return forcing_paths
+
+    @property
+    def _n_time_steps(self) -> int:
+        """Calculate the total number of time steps for the simulation.
+
+        This internal property determines the number of time steps based on the
+        start and end dates of the simulation and the time step size specified in
+        `discretization`.
+
+        Returns
+        -------
+        int
+            The total number of model time steps between `start_date` and `end_date`.
+
+        Raises
+        ------
+        AttributeError
+            If `start_date`, `end_date`, or `discretization.time_step` is not set.
+        """
+
+        run_length_seconds = int((self.end_date - self.start_date).total_seconds())
+        return run_length_seconds // self.discretization.time_step
```

Comment:
This will likely produce an incorrect number of time steps (off by 1). The remainder must be processed in an additional timestep.

add
```py
if run_len_seconds % self.d.ts:
  rls += 1
return rls
```